chore: Bump Snaps packages

Mrtenz opened the pull request:
<!--
Please submit this PR as a draft initially.
Do not mark it as "Ready for review" until the template has been completely filled out, and PR status checks have passed at least once.
-->

## **Description**

This bumps Snaps packages to the latest versions. Notable changes are:

- Support for `wallet_switchEthereumChain` inside Snaps.
   - Like with websites, we want to support a custom network per Snap rather than relying on the globally selected network. For Snaps, the permission for the network to switch to is automatically granted.
- Support for SIP-31 `onClientRequest` handler.
- New `showTestnets` property in `snap_getPreferences`.

[![Open in GitHub Codespaces](https://github.com/codespaces/badge.svg)](https://codespaces.new/MetaMask/metamask-extension/pull/33018?quickstart=1)

## **Related issues**

Closes: https://github.com/MetaMask/metamask-extension/issues/24718

Closes: https://github.com/MetaMask/MetaMask-planning/issues/2940

Closes: https://github.com/MetaMask/MetaMask-planning/issues/2939

## **Manual testing steps**

### `wallet_switchEthereumChain`

1. Go to [`test-snaps`](https://metamask.github.io/snaps/test-snaps/latest/)
2. Install the Ethereum Provider Example Snap.
3. Switch the chain ID using the `test-snaps` selector for the Ethereum Provider example.
4. Click "Get version".
5. The Snap should now use the new network, without any permission requests, and the appropriate chain ID should be shown.

## **Screenshots/Recordings**

<!-- If applicable, add screenshots and/or recordings to visualize the before and after of your change. -->

### **Before**

<!-- [screenshots/recordings] -->

### **After**

<!-- [screenshots/recordings] -->

## **Pre-merge author checklist**

- [ ] I've followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Extension Coding Standards](https://github.com/MetaMask/metamask-extension/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've completed the PR template to the best of my ability
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/main/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

Mrtenz committed:
Bump Snaps packages

Mrtenz commented:
@SocketSecurity ignore npm/@metamask/snaps-execution-environments@8.1.0

This is ours.

Mrtenz committed:
Bump `@metamask/selected-network-controller` to `^22.1.0`

Mrtenz commented:
@metamaskbot update-policies

FrederikBolding committed:
test: Add E2E coverage for initialConnections (#32954)

<!--
Please submit this PR as a draft initially.
Do not mark it as "Ready for review" until the template has been
completely filled out, and PR status checks have passed at least once.
-->

## **Description**

This PR adds E2E coverage for `initialConnections` by using the
preinstalled example Snap which has had `initialConnections`
functionality added and fixes some older E2Es.

[![Open in GitHub
Codespaces](https://github.com/codespaces/badge.svg)](https://codespaces.new/MetaMask/metamask-extension/pull/32954?quickstart=1)

## **Related issues**

Fixes: https://github.com/MetaMask/metamask-extension/issues/24718

GuillaumeRx committed:
feat(snaps): Bind `showTestNetworks` to `snap_getPreferences` hook (#32967)

## **Description**

(Waiting for a Snap platform release)

This PR binds the preference controller `showTestNetworks` property to
the `snap_getPreferences` hook.

[![Open in GitHub
Codespaces](https://github.com/codespaces/badge.svg)](https://codespaces.new/MetaMask/metamask-extension/pull/32967?quickstart=1)

## **Related issues**

Fixes:

## **Manual testing steps**

1. Go to test-snaps
2. Use the preferences snap
3. Verify that `showTestnets` is passed along.

## **Screenshots/Recordings**

<!-- If applicable, add screenshots and/or recordings to visualize the
before and after of your change. -->

### **Before**

<!-- [screenshots/recordings] -->

### **After**


![image](https://github.com/user-attachments/assets/f7498998-ec35-406b-9689-0c41ea518951)

## **Pre-merge author checklist**

- [ ] I've followed [MetaMask Contributor
Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask
Extension Coding
Standards](https://github.com/MetaMask/metamask-extension/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've completed the PR template to the best of my ability
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format
if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling
guidelines](https://github.com/MetaMask/metamask-extension/blob/main/.github/guidelines/LABELING_GUIDELINES.md)).
Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the
app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described
in the ticket it closes and includes the necessary testing evidence such
as recordings and or screenshots.

metamaskbot committed:
Update LavaMoat policies

FrederikBolding commented on package.json:
```diff
@@ -324,14 +324,14 @@
     "@metamask/safe-event-emitter": "^3.1.1",
     "@metamask/scure-bip39": "^2.0.3",
-    "@metamask/selected-network-controller": "^19.0.0",
+    "@metamask/selected-network-controller": "^22.1.0",
```
Do we need to handle any of these breaking changes?

Mrtenz replied:
Doesn't seem like it. 20.0.0 has breaking changes which were reverted in 21.0.0, and the other breaking changes are bumps to `@metamask/network-controller`.

Mrtenz commented:
@metamaskbot update-policies

Mrtenz committed:
Fix lint errors

Mrtenz committed:
Bump `@metamask/preinstalled-example-snap` to `^0.4.0`

Mrtenz committed:
feat(snaps): Add support for custom network per Snap (#26389)

Like with websites, we want to support a custom network per Snap rather
than relying on the globally selected network. For Snaps, the permission
for the network to switch to is automatically granted.

[![Open in GitHub
Codespaces](https://github.com/codespaces/badge.svg)](https://codespaces.new/MetaMask/metamask-extension/pull/26389?quickstart=1)

1. Copy the build files for `@metamask/selected-network-controller` from
[this PR](https://github.com/MetaMask/core/pull/4602).
2. Copy the build files for `@metamask/snaps-controllers` and
`@metamask/snaps-execution-environments` from [this
PR](https://github.com/MetaMask/snaps/pull/2634).
3. Run `test-snaps` for the PR above (`yarn start` in
`packages/test-snaps`).
4. Install the Ethereum Provider Example Snap.
5. Switch the chain ID using the `test-snaps` button for the Ethereum
Provider example.
6. The Snap should now use the new network, without any permission
requests.

Step 1 and 2 are no longer necessary once these PRs are merged and
released.

<!-- If applicable, add screenshots and/or recordings to visualize the
before and after of your change. -->

<!-- [screenshots/recordings] -->

<!-- [screenshots/recordings] -->

- [ ] I've followed [MetaMask Contributor
Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask
Extension Coding
Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've completed the PR template to the best of my ability
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format
if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling
guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)).
Not required for external contributors.

- [ ] I've manually tested the PR (e.g. pull and build branch, run the
app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described
in the ticket it closes and includes the necessary testing evidence such
as recordings and or screenshots.

FrederikBolding commented on yarn.lock:
```diff
@@ -13337,6 +13320,15 @@ __metadata:
+  version: 3.0.0
+  resolution: "@types/nanoid@npm:3.0.0"
+  dependencies:
+    nanoid: "npm:*"
```
Somewhat strange that this brings in another version of nanoid 🤔

FrederikBolding commented on package.json:
```diff
@@ -509,6 +509,7 @@
     "@types/madge": "^5.0.3",
     "@types/micromatch": "^4",
     "@types/mocha": "^10.0.3",
+    "@types/nanoid": "^3.0.0",
```
This doesn't seem to be required? The version we use of `nanoid` everywhere else has types

Mrtenz replied:
You're right. It was required at some point I think, but works fine without it now.

Mrtenz committed:
Remove `@types/nanoid`

Mrtenz committed:
Update example mock

Mrtenz committed:
Fix lint error after rebase

adonesky1 reviewed (approved):
`wallet_switchEthereumChain` enablement changes (including migration) LGTM!

jiexi commented on app/scripts/migrations/160.ts:
```diff
@@ -0,0 +1,314 @@
+    const caip25PermissionCaveatWithCurrentChainIdSet = addPermittedEthChainId(
+      getExistingCaip25PermissionCaveat(subject),
+      currentChainId,
+    );
```
we may want to do a `getEthAccounts` and setEthAccounts` to ensure the existing accounts are set on all scopes? @adonesky1

Mrtenz replied:
Not exactly sure what you mean. Happy to address this in a follow up though!
